@xen-orchestra/fs/src/azure.js

```diff
@@ -0,0 +1,221 @@
+      )
+    } else {
+      info('Connecting to Azure blob storage...')
+      const credentials = new StorageSharedKeyCredential(username, password)
+      this.#blobServiceClient = new BlobServiceClient(`${protocol}://${username}.blob.core.windows.net`, credentials)
+    }
+
+    const parts = split(path)
+    this.#container = parts.shift()
+    this.#dir = join(...parts)
+    this.#containerClient = this.#blobServiceClient.getContainerClient(this.#container)
+    const WITH_RETRY = ['_copy', '_getSize', '_list', '_outputFile', '_read', '_rename', '_unlink', '_writeFile']
+    WITH_RETRY.forEach(functionName => {
+      if (this[functionName] !== undefined) {
+        this[functionName] = pRetry.wrap(this[functionName], {
+          delays: [100, 200, 500, 1000, 2000],
+          when: err => !['SystemInUse'].includes(err?.code),
+          onRetry(error) {
+            warn('retrying method on fs ', {
+              method: functionName,
+              attemptNumber: this.attemptNumber,
+              delay: this.delay,
+              error,
+              file: this.arguments?.[0],
+            })
+          },
+        })
+      }
+    })
+  }
+
+  get type() {
+    return 'azure'
+  }
+  async _sync() {
+    await this.#containerClient.createIfNotExists()
+    await super._sync()
+  }
+
+  async #makePrefix(path) {
+    return path.endsWith('/') ? path : `${path}/`
+  }
+
+  async #isNotEmptyDir(dir) {
+    const prefix = this.#makePrefix(dir)
+    const iterator = this.#containerClient.listBlobsFlat({ prefix }).byPage({ maxPageSize: 1 })
+
+    const { value } = await iterator.next()
+    return (value?.segment?.blobItems?.length ?? 0) > 0
+  }
+
+  async #streamToBuffer(readableStream, buffer) {
+    return new Promise((resolve, reject) => {
+      let bytesRead = 0
+
+      readableStream.on('data', data => {
+        if (bytesRead + data.length <= buffer.length) {
+          data.copy(buffer, bytesRead)
+          bytesRead += data.length
+        } else {
+          reject(new Error('Buffer size exceeded'))
+        }
+      })
+      readableStream.on('end', () => {
+        resolve(bytesRead)
+      })
+      readableStream.on('error', reject)
+    })
+  }
+
+  async _outputStream(file, data) {
+    if (data.length < MIN_BLOCK_SIZE) {
+      await this._writeFile(file, data)
+      return
+    }
+
+    const blobClient = this.#containerClient.getBlockBlobClient(file)
+    const blockCount = Math.ceil(data.length / MAX_BLOCK_SIZE)
+
+    if (blockCount > MAX_BLOCK_COUNT) {
+      throw new Error(`File too large. Maximum upload size is ${MAX_BLOCK_SIZE * MAX_BLOCK_COUNT} bytes`)
+    }
+    const blockIds = []
+
+    for (let i = 0; i < blockCount; i++) {
+      const blockId = Buffer.from(i.toString().padStart(6, '0')).toString('base64')
+      blockIds.push(blockId)
+
+      const start = i * MAX_BLOCK_SIZE
+      const end = Math.min(start + MAX_BLOCK_SIZE, data.length)
+      const chunk = data.slice(start, end)
```

Comment:
data is a stream, you can't use data.slice here. 
You can use readChunkStrict from package @vates/read-chunk to get a part of a stream as a buffer.